src/AbstractSQLCompiler.ts

```diff
@@ -373,64 +373,82 @@ const dataTypeGen = (
 	}
 };
 
-const getReferencedFields: EngineInstance['getReferencedFields'] = ruleBody => {
-	const tableAliases: {
-		[alias: string]: string;
-	} = {};
-	const referencedFields: ReferencedFields = {};
-	const recurse = (rulePart: AbstractSqlQuery) => {
-		_.each(rulePart, part => {
-			if (_.isArray(part)) {
-				if (part[0] === 'ReferencedField') {
-					const [, tableName, fieldName] = part;
-					if (!_.isString(tableName) || !_.isString(fieldName)) {
-						throw new Error('Invalid ReferencedField');
-					}
-					if (referencedFields[tableName] == null) {
-						referencedFields[tableName] = [];
-					}
-					referencedFields[tableName].push(fieldName);
-					return;
-				}
-				if (part[0] === 'Field') {
-					throw new Error('Cannot find queried fields for unreferenced fields');
-				}
-				if (part[0] === 'From') {
-					let nested = (part as FromNode)[1];
-					if (_.isArray(nested)) {
-						if (
-							nested.length === 2 &&
-							nested[0] !== 'Table' &&
-							_.isString(nested[1])
-						) {
-							// Deprecated: Remove this when we drop implicit aliases
-							nested = ['Alias', (nested[0] as any) as TableNode, nested[1]];
-						}
-						if (nested[0] === 'Alias') {
-							let [, table, alias] = nested;
-							if (_.isString(table)) {
-								// Deprecated: Remove this when we drop implicit tables
-								table = ['Table', table];
-							}
-							if (table[0] !== 'Table' || !_.isString(alias)) {
-								throw new Error('Cannot handle aliased select queries');
-							}
-							tableAliases[alias] = table[1];
-						}
-					}
+type Scope = _.Dictionary<string>;
+
+const getScope = (rulePart: AbstractSqlQuery, scope: Scope): Scope => {
+	scope = { ...scope };
+	const fromNodes = rulePart.filter(node => node[0] === 'From') as FromNode[];
+	fromNodes.forEach(node => {
+		let nested = node[1];
+		if (nested[0] === 'Alias') {
+			let [, from, alias] = nested;
+			if (!_.isString(alias)) {
+				throw new Error('Cannot handle non-string aliases');
+			}
+			switch (from[0]) {
+				case 'Table':
+					scope[alias] = from[1];
+					break;
+				case 'SelectQuery':
+					// Ignore SelectQuery in the From as we'll handle any fields it select
+					// when we recurse in.  With true scope handling however we could prune
```

Comment:
extra space

Reply:
Yeah it's a habit of an old school typing teacher, it used to be standard to double space after a period :P